src/main/java/bio/terra/service/filedata/FileIdService.java

```diff
@@ -0,0 +1,26 @@
+package bio.terra.service.filedata;
+
+import bio.terra.service.dataset.Dataset;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.UUID;
+import org.springframework.stereotype.Component;
+
+@Component
+public class FileIdService {
+  private static final String HASH_SEPARATOR = "#";
+
+  public UUID calculateFileId(Dataset dataset, FSItem fsItem) {
+    if (dataset.isPredictableFileIds()) {
+      return UUID.nameUUIDFromBytes(createHashableContent(fsItem).getBytes(StandardCharsets.UTF_8));
+    } else {
+      return UUID.randomUUID();
+    }
+  }
```

Comment:
Right, so the assumption is that if there is a collision, you would only see the file that you have access to (based on the snapshot) so collisions aren't a problem from a security point of view.  Where it might get interesting is if a user has access to both but there's not really a good way to check for collisions and I'm not too concerned given the low probability.  The MD5 (prob = 1.47*10^29) and the path and the size have to match...that probability is exceedingly small